### PR TITLE
Database redesign DRAFT

### LINKED ISSUE #5: Feature: Database Redesign
Currently, our app only allows conversations between two users. We would like to allow conversations with groups of 3 or more users. This ticket is **not** to implement this full change, but to help us plan out what that change would look like. Please make a PR with the changes to just the database models that you would use to implement this feature. Please include "draft" in the title of the PR. Then, in the PR description, write out notes to answer the following two questions:

1) What additional (non-database) changes are needed to implement this feature (we want this to be high level but thorough)?

2) If our app were already deployed, what steps would we need to take to move to this new feature without disrupting service for current users?

Note: In your PR description, please share the thought process that led you to this design.

### PR DESCRIPTION
**Pull request by jjjjonathan**
## Description

- Closes #5 
- Changes database model so that user and conversation models have a many-to-many relationship facilitated by junction table userConversation

## Notes on thought process

Users are associated to conversations via the junction table, “userConversation”, automatically created by Sequelize. Each row in the table consists of a userId and a conversationId. This will allow the table to be used to find all users that are a part of a conversation, and all conversations that a user is a part of.

I chose this approach because it was the recommended approach for many-to-many relationships given in the Sequelize docs.

### What additional (non-database) changes are needed to implement this feature (we want this to be high level but thorough)?

The conversations API route will need to query the userConversation table to find the conversation IDs that match the logged in user’s ID.

Creating a new conversation will involve first creating a new row in the conversation table to get a conversation ID, and then adding the users to the userConversation table with that ID.

The findConversation method has been removed from the Conversation model because conversations will need to be tracked by conversation ID, not by the IDs of the two users.

If a message is trying to be sent to the same group of users, we will need to just send it the existing conversation with that group of users (unless we want to be able to have multiple existing conversations with the same people).

On the frontend, we will need UI to create conversations between multiple users. I would probably rework the search field in the sidebar to only search existing conversations, and add a plus button to add new conversations, which would trigger a menu where multiple users can be selected at once.

### If our app were already deployed, what steps would we need to take to move to this new feature without disrupting service for current users?

To migrate the database, I would write a script that would take the user1Id and user2Id fields from the conversation table and add them to the userConversation table with the corresponding conversation ID.

### CODE REVIEW

**Review by johnsciutto — COMMENTED** [review]
Hi @jjjjonathan

Thank you for your answers 🙂 
I have a few additional questions:

- How would you update the Redux and Socket code in order to adjust for this new feature?
- Could you go into a bit more detail as to how the migration plan would be executed in order to minimize the disruption for users?
- If you had to work on this feature with a team, how would you split the task so that there are no merge conflicts?

When you've answered the new questions, please request a second review with the "/hatchy pr" command one last time 🎉

**Review by sonipranjal — APPROVED** [review]
Hi @jjjjonathan, thank you for answering the questions.

The PR is approved, but don't merge it since it's supposed to be a draft 🙂

**Pull request closed without merging**
